CRM/Mailing/BAO/Mailing.php

```diff
@@ -109,46 +109,227 @@ public static function getRecipientsCount($job_id, $mailing_id = NULL) {
+   *  1. Takes mailing ID as the only parameter, and doesn't depend on mailing Job
+   *  2. This function ONLY fetch and store mailing recipients, doesn't handle SMS mode
+   *  3. Use less JOINs in underlying queries.
+   *  4. Simplifies code to fetch contacts from selected mailing smart group(s).
+   *  5. Doesn't execute query needlessly, e.g.
+   *    a) if SG (smart groups) are selected then ONLY queries resposible to fetch contacts from SG are executed
+   *    b) If no 'Include' groups are selected then simply return as this means no groups are selected to fetch recipients.
+   *  6. Refreshes smart group contact cache at once.
+   *
+   * @param CRM_Mailing_DAO_Mailing $mailingObj
    *
-   * @return CRM_Mailing_Event_BAO_Queue|string
+   * @return void
    */
-  public static function getRecipients(
-    $job_id,
-    $mailing_id = NULL,
-    $storeRecipients = FALSE,
-    $dedupeEmail = FALSE,
-    $mode = NULL) {
+  public static function getMailRecipients($mailingObj) {
+    $mailing = CRM_Mailing_BAO_Mailing::getTableName();
+    $mailingID = $mailingObj->id;
+
     $mailingGroup = new CRM_Mailing_DAO_MailingGroup();
+    $recipientsGroup = $excludeSmartGroupIDs = $includeSmartGroupIDs = array();
+    $sql = " SELECT GROUP_CONCAT(entity_id SEPARATOR \",\") as group_ids, group_type
+      FROM civicrm_mailing_group
+      WHERE mailing_id = %1 AND entity_table = 'civicrm_group'
+      GROUP BY group_type
+    ";
+    $dao = CRM_Core_DAO::executeQuery($sql, array(1 => array($mailingID, 'Int')));
+    while ($dao->fetch()) {
+      $recipientsGroup[$dao->group_type] = explode(',', $dao->group_ids);
+    }
 
-    $mailing = CRM_Mailing_BAO_Mailing::getTableName();
-    $job = CRM_Mailing_BAO_MailingJob::getTableName();
-    $mg = CRM_Mailing_DAO_MailingGroup::getTableName();
-    $eq = CRM_Mailing_Event_DAO_Queue::getTableName();
+    // there is no need to proceed further as no mailing group is selected to include recipients,
+    // but before return clear the mailing recipients populated earlier since as per current params no group is selected
+    if (empty($recipientsGroup['Include'])) {
+      $sql = " DELETE FROM civicrm_mailing_recipients WHERE  mailing_id = %1 ";
+      CRM_Core_DAO::executeQuery($sql, array(1 => array($mailingID, 'Integer')));
+      return;
+    }
+
+    list($location_filter, $order_by) = self::getLocationFilterAndOrderBy($mailingObj->email_selection_method, $mailingObj->location_type_id);
+
+    // get all the saved searches AND hierarchical groups
+    // and load them in the cache
+    foreach ($recipientsGroup as $groupType => $groupIDs) {
+      $sql = sprintf("
+       SELECT *
+       FROM  civicrm_group g
+       WHERE id IN (%s) AND ( saved_search_id != 0 OR saved_search_id IS NOT NULL OR children IS NOT NULL )
+      ", implode(',', $groupIDs));
+
+      $groupDAO = CRM_Core_DAO::executeQuery($sql);
+      while ($groupDAO->fetch()) {
+        if ($groupDAO->cache_date == NULL) {
```

Comment:
It seems like you are not rebuilding the cache for *all* groups, just for ones where the cache is empty/expired? If so, please clarify the DocBlock for the function.

Reply:
Yes if cache is empty as per earlier code in ```getRecipients(...)``, I haven't changed this logic. Will mention it in doc block